# CLASE 19: CÓMO CALCULAR DISTANCIAS ENTRE VECTORES

# 8. Distancia y Norma

En clases pasadas hablamos de la magnitud de un vector pero no dimos una forma de poder medirla, esa magnitud la podremos calcular a través de la *norma Euclidiana*:

$$
||x|| = \displaystyle\sqrt{x_{0}^2 + x_{1}^2 + \cdots + x_{n-1}^{2}} = \displaystyle\sqrt(x^Tx)
$$

Hay que notar que $x$ en la expresión anterior es un $n$-vector. 
Pongamos un ejemplo con $x^T = [1,2,3]$

$$
||x|| = \displaystyle\sqrt{1^2 + 2^2 + 3^2} = \sqrt{14} \approx 3.74
$$

Realizaremos otro con $x^T = [2,-1,2]$

$$
||x|| = \displaystyle\sqrt{2^2 + (-1)^2 + 2^2} = \sqrt{9} = 3
$$

## 8.1 Propieades de la norma

Como todas las operaciones con las que trabajamos hasta ahora la norma también cumple una serie de propiedades. Sean $x$, $y$ son $n$-vectores y $\beta$ un escalar cualquiera, tenemos que la norma cumple:
* Homogeneídad no negativa: $||\beta x|| = |\beta|\cdot ||x||$

* Desigualdad del triángulo: $||x+y|| \le ||x|| + ||y||$ 

* No-negatividad: $||x||\ge 0$ y además $||x|| = 0$ si y solamente si $x = \mathbf{0}$

La norma es ampliamente usada, en particular podemos señalar los siguientes ejemplos:

**Valor cuadrático medio.** Algunas veces también es llamado simplemente **RMS** (root mean square) y se define como sigue:

$$
rms(x) = \displaystyle\sqrt{\frac{x_{0}^2 + x_{1}^2 + \cdots + x_{n-1}^{2}}{n}} = \frac{||x||}{\sqrt{n}}
$$

**Norma de una suma.** Si tenemos dos vectores $x$ y $y$ entonces la norma de la suma es

$$
||x+y|| = \displaystyle\sqrt{||x||^{2} + 2x^{T}y + ||y||^{2}}
$$

**Norma de una bloque de vectores.** La norma de bloque de vectores será simplemente la raíz cuadrada de la suma de sus componentes. Si tenemos $d$ un $k$-vector con componentes $d_{i}$ que son $n$-vectores entonces tenemos que:

$$
||d|| = \displaystyle\sqrt{||d_{0}||^{2}+||d_{1}||^{2}+\cdots + ||d_{n-1}||^{2}}
$$

In [1]:
# Se importa la librerías necesarias
import numpy as np

In [2]:
# Función para hallar la norma de un vectos
def norma(x):
    return np.sqrt(x@x)

In [6]:
# Se crea ek vector
u = np.array([1,1])

# Se halla la norma
print('norma(u):\n', norma(u))
print('square_root(2):\n', np.sqrt(2))

norma(u):
 1.4142135623730951
square_root(2):
 1.4142135623730951


Efectivamente la norma halla el valor de la raiz cuadrada de los elementos al cuadrado.

Ahora se comprueba la propiedad de la homogeneidad no negativa con el vector u y un escalar de 2.

$2*u^T = [2,2]$ entonces la norma será $||2u|| = \displaystyle\sqrt{2^2 + 2^2} = \displaystyle\sqrt{8}$

In [8]:
# Se comprueba la propiedad de la homogeneidad no negativa
print('norma(2u):\n', norma(2*u))
print('2*norma(u):\n', 2*norma(u))

norma(2u):
 2.8284271247461903
2*norma(u):
 2.8284271247461903


De esta forma se comprueba la propiedad de la homogeneidad no negativa. Ahora se comprueba la propiedad de la desigualdad del triangulo, el cual esta dada por:

* Desigualdad del triángulo: $||x+y|| \le ||x|| + ||y||$ 

In [9]:
# Se crea un nuevo vector
v = np.array([-1,1])
# Se comprueba la propiedad
print('u:\n', u)
print('v:\n', v)
print('||u + v||:\n', norma(u + v))
print('¿Es verdad que ||u + v|| <= ||u|| + ||v||?: ', norma(u+v) <= norma(u) + norma(v))

u:
 [1 1]
v:
 [-1  1]
||u + v||:
 2.0
¿Es verdad que ||u + v|| <= ||u|| + ||v||?:  True


Se comprueba la propiedad de la desigualdad triangular.

# CLASE 20: DISTANCIA ENTRE VECTORES

## 8.1 Distancia entre vectores

Podemos usar la norma pora definir la distancia entre dos vectores $\mathbf{a}$ y $\mathbf{b}$ como la norma de la diferencia de los vectores

$$
dist(\mathbf{a},\mathbf{b}) = ||\mathbf{a} - \mathbf{b}|| = \displaystyle\sqrt{(a_{0}-b_{0})^2 + (a_{1}-b_{1})^2 + \cdots + (a_{n-1}-b_{n-1})^2}
$$

In [10]:
# Se escriben dos vectores
a = np.array([1,1])
b = np.array([-2,3])

# Ahora se halla la resta de los dos vectores
c = a - b

print('a:\n', a)
print('b:\n', b)
print('c = a - b:\n', c)

a:
 [1 1]
b:
 [-2  3]
c = a - b:
 [ 3 -2]


Recordemos que sucede cuando se suman dos vectores

![](imagenes/img7.png)

El vector w es la suma de los vectores v + u, el cuál se puede formar por medio de la suma del paralelogramo (vectores en rojo)

![](imagenes/img8.png)

Ahora, si en ves de sumar a u le restamos el vector v, nos da como resultado el vector c, el cuál si lo trasladamos hasta el final del vector v, este nuevo vector d termina en donde termina el vector u, port tanto la norma de la resta de dos vectores da como resultado la distancia entre los dos vectores. 

In [12]:
# Distancia entre dos vectores
print('distancia:\n', norma(a - b))

distancia:
 3.605551275463989


Sin embargo numpy tiene la función para hallar la norma de un vector

In [13]:
# Función de python para hallar la norma
print('np.linalg.norm(a - b):\n', np.linalg.norm(a - b))

np.linalg.norm(a - b):
 3.605551275463989


A esto se el llama _norma Euclidiana_ o *distancia Euclidiana* y tiene ese nombre porque es como se infiere la forma de medir distancias de los [postulados de Euclides](https://es.wikipedia.org/wiki/Postulados_de_Euclides). Ahora, la realidad es que siempre que nuestro espacio admita una $\mathbf{L^p}$-norma entonces podremos generalizar la norma anterior como

$$
||x||_{p} = \left(|x_{0}|^p + |x_{1}|^p + \cdots + |x_{n-1}|^p\right)^{1/p} \qquad con \;\; 1\le p
$$

Por eso es que la norma Euclidiana también es conocida como la *norma $L_{2}$*. Hay otras normas que recibe nombres. Por ejemplo, si $p=1$ tenemos la norma $L_{1}$ y esta es llamada _la métrica del taxista_ o _la métrica de Manhattan_. Si tomamos $p \to \infty$ entonces tenemos $L_{\infty}$ y esta es llamada _la métrica del máximo_. Es muy importante saber qué métrica estamos usando para medir distancias ya que la forma en la que medimos distancias está en total correspondencia en con configuración y la percepción el espacio. Por ejemplo, mientras que en la métrica Euclidiana la distancia corresponde a la más corta entre los dos vectores en el espacio planeo en el caso de la métrica del taxista es la más corta pero no será es segmento más directo sino por _bloques_.

Para poder simplificar las cosas tomemos dos vectores $\mathbf{x}$ y $\mathbf{y}$ en $\mathbb{R}^2$

$$
\mathbf{x} = \begin{bmatrix} 2\\1 \end{bmatrix},\; \qquad \mathbf{y} = \begin{bmatrix} 1\\2 \end{bmatrix}
$$

La distancia en el caso de norma $L_{2}$ tenemos que la distancia entre los vectores es $||\mathbf{x}-\mathbf{y}||_{2} = \displaystyle\sqrt{(2-1)^{2} + (1-2)^{2}} = \displaystyle\sqrt{2}$ mientras que bajo la norma $L_{1}$ obtenemos que $||\mathbf{x}-\mathbf{y}||_{1} = |2-1|+|1-2| = 2$. 

Esto se puede representar graficamente como:

![](imagenes/img9.png)

Donde los vectores son u, v, la norma L2 esta representado en rojo mientras que la norma L1 esta representado en azul.

Ahora, si $\mathbf{x}$ y $\mathbf{y}$ son $n$-vectores hay una medida quer después cobrará importancia, *la desviación cuadrática media*

$$
rms_{dev}(\mathbf{x},\mathbf{y}) = \displaystyle\frac{||\mathbf{x}-\mathbf{y}||}{\sqrt{n}}
$$

Algo muy imporante es que a pesar de que medimos distancias no existen propiamente nociones de *cerca* y *lejos*. Si tomamos los siguientes $4$-vectores

$$
\mathbf{u} = \begin{bmatrix}1.8 \\ 2.8 \\ -3.7 \\ 4.7 \end{bmatrix}, \qquad  \mathbf{v} = \begin{bmatrix}0.6 \\ 2.1 \\ 1.9 \\ -1.4 \end{bmatrix}, \qquad \mathbf{w} = \begin{bmatrix}2.0 \\ 1.9 \\ -4.0 \\ 4.6 \end{bmatrix}
$$

Tenemos que las distancia entre los vectores es:

$$
||\mathbf{u}-\mathbf{v}|| = 8.36, \qquad ||\mathbf{u}-\mathbf{w}|| = 0.38, \qquad ||\mathbf{v}-\mathbf{w}|| = 8.63
$$

Entonces así ya tenemos noción de cercanía o lontanaza, de no ser porque comparamos no tendríamos estas nociones. 

**Desigualdad del triángulo.** Debido a que la distancia la definimos a partir de la norma entonces la distancia también está sujeta a cumplir la desigualdad del triángulo. Considemos entonces tres $n$-vectores $\mathbf{a}$,$\mathbf{b}$ y $\mathbf{c}$ entonces tenemos

$$
||\mathbf{a}-\mathbf{c}|| = ||\mathbf{a}-\mathbf{b}+\mathbf{b}-\mathbf{c}|| \le ||\mathbf{a}-\mathbf{b}|| + ||\mathbf{b}-\mathbf{c}||
$$

Vamos a partir de los siguientes $2$-vectores:
$$
\mathbf{a} = \begin{bmatrix} 1\\ 1 \end{bmatrix}, \qquad \mathbf{b} = \begin{bmatrix} 2\\ 3 \end{bmatrix}, \qquad \mathbf{c} = \begin{bmatrix} 4\\ 2 \end{bmatrix}
$$

## Enlaces Externos

[Postulados de Euclides](https://es.wikipedia.org/wiki/Postulados_de_Euclides)

[Colab Original](https://colab.research.google.com/drive/1pg6fxVEmEDlpou3JbEtnlZzcVckPGS8F?usp=sharing)